File: pcces-go/internal/storage/sqlite/cost_structure_project_run.go
package sqlite

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	errx "github.com/liuxb99/pcces-web/pcces-go/internal/platform/errors"
	"github.com/shopspring/decimal"
)

type BudgetSnapshotItem struct {
	ID        string `json:"id"`
	Kind      string `json:"kind"`
	Quantity  string `json:"quantity"`
	UnitPrice string `json:"unit_price"`
	Amount    string `json:"amount"`
}

type ProjectCostStructureRun struct {
	ID                  string                `json:"id"`
	ProjectCode         string                `json:"project_code"`
	CostStructureTypeID string                `json:"cost_structure_type_id"`
	DirectCost          string                `json:"direct_cost"`
	Total               string                `json:"total"`
	Scale               int32                 `json:"scale"`
	BudgetSnapshot      []BudgetSnapshotItem  `json:"budget_snapshot"`
	Result              CostCalculationResult `json:"result"`
	CreatedBy           string                `json:"created_by"`
	CreatedAt           string                `json:"created_at"`
	RowVersion          int64                 `json:"row_version"`
	DeepLink            string                `json:"deep_link"`
}

type ProjectCostStructureRunRepository struct{ store *Store }

func NewProjectCostStructureRunRepository(store *Store) *ProjectCostStructureRunRepository {
	return &ProjectCostStructureRunRepository{store: store}
}

func aggregateBudget(items []BudgetSnapshotItem) (decimal.Decimal, error) {
	total := decimal.Zero
	for _, item := range items {
		kind := strings.ToUpper(strings.TrimSpace(item.Kind))
		if kind == "SECTION" || kind == "CHAPTER" || kind == "FOLDER" {
			continue
		}
		var amount decimal.Decimal
		var err error
		if strings.TrimSpace(item.Amount) != "" {
			amount, err = decimal.NewFromString(item.Amount)
		} else {
			qty, qtyErr := decimal.NewFromString(defaultString(item.Quantity, "0"))
			if qtyErr != nil {
				return decimal.Zero, qtyErr
			}
			price, priceErr := decimal.NewFromString(defaultString(item.UnitPrice, "0"))
			if priceErr != nil {
				return decimal.Zero, priceErr
			}
			amount = qty.Mul(price)
		}
		if err != nil {
			return decimal.Zero, err
		}
		total = total.Add(amount)
	}
	return total, nil
}

func (r *ProjectCostStructureRunRepository) Recalculate(ctx context.Context, projectCode string, items []BudgetSnapshotItem, scale int32, actor string) (ProjectCostStructureRun, error) {
	projectCode = strings.TrimSpace(projectCode)
	if projectCode == "" {
		return ProjectCostStructureRun{}, errx.New(errx.CodeInvalidArgument, "project_code is required", "P4-COST-006")
	}
	var typeID string
	if err := r.store.db.QueryRowContext(ctx, `SELECT cost_structure_type_id FROM project_cost_structures WHERE project_code=?`, projectCode).Scan(&typeID); err == sql.ErrNoRows {
		return ProjectCostStructureRun{}, errx.New(errx.CodeNotFound, "project cost structure not assigned", "P4-COST-006")
	} else if err != nil {
		return ProjectCostStructureRun{}, err
	}
	rows, err := r.store.db.QueryContext(ctx, `SELECT code,kind,rate,sequence FROM cost_structure_categories WHERE cost_structure_type_id=? AND enabled=1 ORDER BY sequence,code`, typeID)
	if err != nil {
		return ProjectCostStructureRun{}, err
	}
	defer rows.Close()
	lines := []CostCalculationLine{}
	for rows.Next() {
		var line CostCalculationLine
		if err := rows.Scan(&line.Code, &line.Kind, &line.Rate, &line.SortOrder); err != nil {
			return ProjectCostStructureRun{}, err
		}
		line.BaseKind = "SUBTOTAL"
		line.Sign = 1
		lines = append(lines, line)
	}
	if len(lines) == 0 {
		return ProjectCostStructureRun{}, errx.New(errx.CodeInvalidArgument, "assigned cost structure has no enabled categories", "P4-COST-006")
	}
	direct, err := aggregateBudget(items)
	if err != nil {
		return ProjectCostStructureRun{}, errx.Wrap(errx.CodeInvalidArgument, "invalid budget amount", "P4-COST-006", err)
	}
	result, err := CalculateCostStructure(lines, direct.String(), scale)
	if err != nil {
		return ProjectCostStructureRun{}, errx.Wrap(errx.CodeInvalidArgument, "calculate cost structure", "P4-COST-006", err)
	}
	budgetJSON, _ := json.Marshal(items)
	resultJSON, _ := json.Marshal(result)
	runID := fmt.Sprintf("CSR-%d", time.Now().UTC().UnixNano())
	now := time.Now().UTC().Format(time.RFC3339Nano)
	_, err = r.store.db.ExecContext(ctx, `INSERT INTO project_cost_structure_runs(id,project_code,cost_structure_type_id,direct_cost,total,scale,budget_snapshot_json,result_json,created_by,created_at,row_version) VALUES(?,?,?,?,?,?,?,?,?,?,1)`, runID, projectCode, typeID, result.DirectCost, result.Total, scale, string(budgetJSON), string(resultJSON), actor, now)
	if err != nil {
		return ProjectCostStructureRun{}, err
	}
	return r.Get(ctx, runID)
}

func (r *ProjectCostStructureRunRepository) Get(ctx context.Context, id string) (ProjectCostStructureRun, error) {
	var item ProjectCostStructureRun
	var budgetJSON, resultJSON string
	err := r.store.db.QueryRowContext(ctx, `SELECT id,project_code,cost_structure_type_id,direct_cost,total,scale,budget_snapshot_json,result_json,created_by,created_at,row_version FROM project_cost_structure_runs WHERE id=?`, id).Scan(&item.ID, &item.ProjectCode, &item.CostStructureTypeID, &item.DirectCost, &item.Total, &item.Scale, &budgetJSON, &resultJSON, &item.CreatedBy, &item.CreatedAt, &item.RowVersion)
	if err == sql.ErrNoRows {
		return item, errx.New(errx.CodeNotFound, "project cost structure run not found", "P4-COST-006")
	}
	if err != nil {
		return item, err
	}
	if err = json.Unmarshal([]byte(budgetJSON), &item.BudgetSnapshot); err != nil {
		return item, err
	}
	if err = json.Unmarshal([]byte(resultJSON), &item.Result); err != nil {
		return item, err
	}
	item.DeepLink = "/app/cost-structure?project=" + item.ProjectCode + "&run=" + item.ID
	return item, nil
}

File: api/cost_structure_project_run.py
"""Persisted project cost-structure initialization and recalculation."""
from __future__ import annotations

import json
from datetime import datetime, timezone
from decimal import Decimal
from uuid import uuid4

from flask import Blueprint, jsonify, request
from sqlalchemy import Column, DateTime, Integer, MetaData, Numeric, String, Table, Text, and_, select

from api.cost_structure_calculation import calculate_cost_structure
from api.cost_structure_details import cost_structure_categories
from api.cost_structure import project_cost_structures
from api.decimal_math import quantize

metadata = MetaData()
project_cost_structure_runs = Table(
    "project_cost_structure_runs", metadata,
    Column("id", String(100), primary_key=True),
    Column("project_code", String(100), nullable=False, index=True),
    Column("cost_structure_type_id", String(100), nullable=False),
    Column("direct_cost", Numeric(24, 8), nullable=False),
    Column("total", Numeric(24, 8), nullable=False),
    Column("scale", Integer, nullable=False),
    Column("budget_snapshot_json", Text, nullable=False),
    Column("result_json", Text, nullable=False),
    Column("created_by", String(100), nullable=False),
    Column("created_at", DateTime(timezone=True), nullable=False),
    Column("row_version", Integer, nullable=False, default=1),
)


class ProjectCostStructureRunService:
    def __init__(self, engine):
        self.engine = engine
        metadata.create_all(engine)

    @staticmethod
    def _direct_cost(items: list[dict]) -> Decimal:
        total = Decimal("0")
        for item in items:
            if str(item.get("kind", "WORK")).upper() in {"SECTION", "CHAPTER", "FOLDER"}:
                continue
            amount = item.get("amount")
            if amount is None:
                amount = Decimal(str(item.get("quantity", "0"))) * Decimal(str(item.get("unit_price", "0")))
            total += Decimal(str(amount))
        return total

    def recalculate(self, project_code: str, budget_items: list[dict], scale: int, actor: str) -> dict:
        if not project_code.strip():
            raise ValueError("project_code is required")
        with self.engine.begin() as conn:
            assignment = conn.execute(select(project_cost_structures).where(
                project_cost_structures.c.project_code == project_code
            )).mappings().first()
            if not assignment:
                raise LookupError("project cost structure not assigned")
            rows = conn.execute(select(cost_structure_categories).where(and_(
                cost_structure_categories.c.cost_structure_type_id == assignment["cost_structure_type_id"],
                cost_structure_categories.c.enabled.is_(True),
            )).order_by(cost_structure_categories.c.sequence, cost_structure_categories.c.code)).mappings().all()
            if not rows:
                raise ValueError("assigned cost structure has no enabled categories")
            lines = [{
                "code": row["code"], "kind": row["kind"], "base_kind": "SUBTOTAL",
                "rate": str(row["rate"]), "sign": 1, "sort_order": row["sequence"],
            } for row in rows]
            direct = self._direct_cost(budget_items)
            result = calculate_cost_structure(lines, str(direct), scale)
            now = datetime.now(timezone.utc)
            run_id = str(uuid4())
            conn.execute(project_cost_structure_runs.insert().values(
                id=run_id, project_code=project_code,
                cost_structure_type_id=assignment["cost_structure_type_id"],
                direct_cost=Decimal(result["direct_cost"]), total=Decimal(result["total"]), scale=scale,
                budget_snapshot_json=json.dumps(budget_items, ensure_ascii=False, sort_keys=True),
                result_json=json.dumps(result, ensure_ascii=False, sort_keys=True),
                created_by=actor, created_at=now, row_version=1,
            ))
        return self.get(run_id)

    def get(self, run_id: str) -> dict:
        with self.engine.connect() as conn:
            row = conn.execute(select(project_cost_structure_runs).where(project_cost_structure_runs.c.id == run_id)).mappings().first()
        if not row:
            raise LookupError("project cost structure run not found")
        return {
            "id": row["id"], "project_code": row["project_code"],
            "cost_structure_type_id": row["cost_structure_type_id"],
            "direct_cost": quantize(str(row["direct_cost"]), row["scale"]), "total": quantize(str(row["total"]), row["scale"]),
            "scale": row["scale"], "budget_snapshot": json.loads(row["budget_snapshot_json"]),
            "result": json.loads(row["result_json"]), "created_by": row["created_by"],
            "created_at": row["created_at"].isoformat(), "row_version": row["row_version"],
            "deep_link": f"/app/cost-structure?project={row['project_code']}&run={row['id']}",
        }


def build_project_cost_structure_run_blueprint(service: ProjectCostStructureRunService, resolve_user_id):
    bp = Blueprint("project_cost_structure_run", __name__, url_prefix="/api/cost-structures")

    @bp.post("/projects/<project_code>/recalculate")
    def recalculate(project_code: str):
        actor = resolve_user_id()
        if actor is None:
            return jsonify({"code": "UNAUTHORIZED"}), 401
        body = request.get_json(silent=True) or {}
        try:
            return jsonify(service.recalculate(project_code, list(body.get("budget_items") or []), int(body.get("scale", 2)), str(actor)))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404
        except (ValueError, ArithmeticError) as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400

    @bp.get("/runs/<run_id>")
    def get_run(run_id: str):
        try:
            return jsonify(service.get(run_id))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404

    return bp
